tests/subsys/net/lib/nrf_cloud/fota_common/src/main.c

```diff
@@ -0,0 +1,716 @@
+#include "fakes.h"
+
+/* This function runs before each test */
+static void run_before(void *fixture)
+{
+	ARG_UNUSED(fixture);
+}
+
+/* This function runs after each completed test */
+static void run_after(void *fixture)
+{
+	ARG_UNUSED(fixture);
+}
+
+ZTEST_SUITE(nrf_cloud_fota_common_test, NULL, NULL, run_before, run_after, NULL);
+
+ZTEST(nrf_cloud_fota_common_test, test_nrf_cloud_fota_is_type_modem)
+{
+	/* Test range beyond enum values */
+	for (enum nrf_cloud_fota_type fota_type = NRF_CLOUD_FOTA_TYPE__FIRST - 1;
+	     fota_type <= NRF_CLOUD_FOTA_TYPE__INVALID + 1;
+	     ++fota_type) {
+
+		if ((fota_type == NRF_CLOUD_FOTA_MODEM_FULL) ||
+		    (fota_type == NRF_CLOUD_FOTA_MODEM_DELTA)) {
+			zassert_true(nrf_cloud_fota_is_type_modem(fota_type),
+				     "Modem FOTA type not detected");
+		} else {
+			zassert_false(nrf_cloud_fota_is_type_modem(fota_type),
+				      "Non-modem FOTA type detected as modem type");
+		}
+	}
+}
+
+#define APP_FOTA_ASSERT_STR "APP FOTA enabled should correspond to the value of: " \
+			    "CONFIG_BOOTLOADER_MCUBOOT"
+
+#define BOOT_FOTA_ASSERT_STR "BOOT FOTA enabled should correspond to the value of: " \
+			     "CONFIG_BOOTLOADER_MCUBOOT + CONFIG_BUILD_S1_VARIANT + " \
+			     "CONFIG_SECURE_BOOT"
+
+#define MDM_DELTA_ASSERT_STR "DELTA MODEM FOTA enabled should correspond to the value of: " \
+			     "CONFIG_NRF_MODEM"
+
+#define FMFU_ASSERT_STR "FULL MODEM FOTA enabled should correspond to the value of: " \
+			"CONFIG_NRF_CLOUD_FOTA_FULL_MODEM_UPDATE + CONFIG_NRF_MODEM"
+
+ZTEST(nrf_cloud_fota_common_test, test_nrf_cloud_fota_is_type_enabled)
+{
+	Z_TEST_SKIP_IFNDEF(CONFIG_FOTA_DOWNLOAD);
+
+	bool ret;
+
+	ret = nrf_cloud_fota_is_type_enabled(NRF_CLOUD_FOTA_APPLICATION);
+	zassert_equal(IS_ENABLED(CONFIG_BOOTLOADER_MCUBOOT), ret, APP_FOTA_ASSERT_STR);
+
+	ret = nrf_cloud_fota_is_type_enabled(NRF_CLOUD_FOTA_BOOTLOADER);
+	zassert_equal(IS_ENABLED(CONFIG_BOOTLOADER_MCUBOOT) &&
+		      IS_ENABLED(CONFIG_BUILD_S1_VARIANT) &&
+		      IS_ENABLED(CONFIG_SECURE_BOOT),
+		      ret,
+		      BOOT_FOTA_ASSERT_STR);
+
+	ret = nrf_cloud_fota_is_type_enabled(NRF_CLOUD_FOTA_MODEM_DELTA);
+	zassert_equal(IS_ENABLED(CONFIG_NRF_MODEM), ret, MDM_DELTA_ASSERT_STR);
+
+	ret = nrf_cloud_fota_is_type_enabled(NRF_CLOUD_FOTA_MODEM_FULL);
+	zassert_equal(IS_ENABLED(CONFIG_NRF_CLOUD_FOTA_FULL_MODEM_UPDATE) &&
+		      IS_ENABLED(CONFIG_NRF_MODEM),
+		      ret,
+		      FMFU_ASSERT_STR);
+
+	ret = nrf_cloud_fota_is_type_enabled(NRF_CLOUD_FOTA_TYPE__INVALID);
+	zassert_false(ret, "return should be false when invalid FOTA type is provided");
+}
+
+ZTEST(nrf_cloud_fota_common_test, test_nrf_cloud_bootloader_fota_slot_set_get_s0_get_fail)
+{
+	Z_TEST_SKIP_IFNDEF(CONFIG_FOTA_DOWNLOAD);
+	struct nrf_cloud_settings_fota_job job;
+
+	job.bl_flags = NRF_CLOUD_FOTA_BL_STATUS_CLEAR;
+	job.type = NRF_CLOUD_FOTA_BOOTLOADER;
+
+	fota_download_s0_active_get_fake.custom_fake = fake_fota_download_s0_active_get__fails;
```

Comment:
Does FFF automatically reset custom fakes at the start of each test?

Reply:
It seems possibly not -- and if that is the case, we will need to explicitly do so

Calling `RESET_FAKE(fota_download_s0_active_get);` and similar for all fakes from the `run_before` function would suffice. But definitely double check that this is actually needed

Reply:
> does simply setting the fake function to the desired value before use have some unintended consequence?

The main fear is this: If you, or anyone after you, ever accidentally writes a test which depends on which fake has been loaded, and forgets to explicitly set the fake in the test, then that test will have differing behavior depending on the order in which tests are executed

Such a test is very undesirable because it may appear fine until some subtle change is made in the tests that run before it, and then it may suddenly fail

So generally it is desirable to do what you can to guarantee that the starting conditions for each test is totally independent of what happened in the tests that ran before it. And so you would want to reset your fakes between each test

Reply:
so, no unintended consequences... just good practice to reset the fakes ✔️ 

looks like other projects do the following in setup:
```
/* Register resets */
DO_FOREACH_FAKE(RESET_FAKE);
/* reset common FFF internal structures */
FFF_RESET_HISTORY();
```
so i'll add that to `run_before `.

Reply:
My bad, I was kind of conceiving of it as an unintended consequence, but it is more of a good practice thing yeah

I think that `DO_FOREACH_FAKE` thing is specific to the modem team's unit tests, they had to actually write a custom macro for that